### ISSUE TITLE
EQL: "Arrays are not supported" error in "sequence by" statements

### ISSUE THREAD

**Issue by peluja1012** [elastic/elasticsearch#64553]
**Description of the problem including expected versus actual behavior**:
EQL queries with "sequence by" statements fail to execute with an error when any of the documents in a sequence contain an array value in the aggregation field.

**Elasticsearch version** (`bin/elasticsearch --version`):
```js
  "version" : {
    "number" : "8.0.0-SNAPSHOT",
    "build_flavor" : "default",
    "build_type" : "deb",
    "build_hash" : "8f8d2d1b4ba31c6da040598a91e9aa559e093a4a",
  }
```

**Steps to reproduce**:
 1. Index two documents. The first document should have a single value for `host.name`. The second document should have an array of values for `host.name`:
```js
PUT /test-index-1/_doc/1
{
  "@timestamp": "2020-11-03T08:51:55.337Z",
  "host": {
    "name": "testhost"
  }
}

PUT /test-index-1/_doc/2
{
  "@timestamp": "2020-11-03T08:51:56.337Z",
  "host": {
    "name": ["testhost", "foo"]
  }
}
```
 2. Run an EQL sequence query with a `by` statement like so:
```js
GET /test-index-1/_eql/search
{
  "query": """
    sequence by host.name with maxspan=15m
    [any where true]
    [any where true]
  """
}
```
 3. Observe that the following error is returned:
```js
{
  "error" : {
    "root_cause" : [
      {
        "type" : "ql_illegal_argument_exception",
        "reason" : "Arrays (returned by [host.name]) are not supported"
      }
    ],
    "type" : "ql_illegal_argument_exception",
    "reason" : "Arrays (returned by [host.name]) are not supported"
  },
  "status" : 500
}
```

**Comment by rw-access** [comment]
@peluja1012 what's your expected behavior? I'm not totally sure what better path we could take.

It's not possible to join with `by` on an array or an object, but only a list of primitives (`bool`, `number`, `string`). This just comes with the territory of joins and EQL. If EQL encounters an array when joining, what's your proposed behavior?  Silently ignore the message, raise an error, or something else?

Also FYI array fields really aren't supported ([link to the docs](https://www.elastic.co/guide/en/elasticsearch/reference/master/eql-syntax.html#eql-array-fields)). We could still leave issue open as a design question, but I think technically it's not a bug?

**Comment by peluja1012** [comment]
I'm definitely ok with removing the "bug" label and keeping this as discussion issue. One thing I noticed is that if there are multiple sequence hits and all of them are valid except for one, the query fails. I wonder if it would be better to have it return the sequences that were valid and add some kind of message in the metadata of the query results noting that x number of sequences were skipped due to errors. Our concern is that Security app users tend to use array values for ECS fields quite often and they probably expect their EQL rules to work when joining by these fields. We could certainly highlight this behavior in the rule documentation but I'm trying to see if there is any way we could improve this behavior.

**Comment by SHolzhauer** [comment]
With a recent case I ran into this as well;

My case:
1. ML triggers on a rare `dns.question.name` for an entities history.
2. There is traffic to one of the `dns.resolved_ip`'s.

The issue here is that `dns.resolved_ip` is an array of ip addresses and `destination.ip` is a keyword.
Making it currently not possible(for as far as I can see) to build this use case.

In [discussion on slack](https://elasticstack.slack.com/archives/C016E72DWDS/p1617970216028700) with @rw-access we did come to the conclusion two detection we're needed but you then run into this issue.
